### PR TITLE
Implement alternate branch root files for #4691

### PR DESCRIPTION
**Pull request by evankanderson**
Fixes https://github.com/gohugoio/hugo/issues/4691

This is a different approach than the first time.

I still need to write up the docs, but there is an e2e test. There are also several tests where the full `FileMetaInfo` isn't present in the filesystem -- what's your preference on how those go in or get cleaned up?

### CODE REVIEW

File: create/content_template_handler.go [hunk 1]
```diff
@@ -97,8 +97,7 @@ func executeArcheTypeAsTemplate(s *hugolib.Site, name, kind, targetPath, archety
 
 	if name == "" {
 		name = f.TranslationBaseName()
-
-		if name == "index" || name == "_index" {
+		if f.FileInfo().Meta().Classifier().IsBundle() || name == "index" {
```
Comment (evankanderson) [inline create/content_template_handler.go:100]: There are 3 test failures if I remove the "index" test here: `TestNewContent/lang-11` through `TestNewContent/lang-13`.

File: deps/deps.go [hunk 2]
```diff
@@ -87,9 +87,6 @@ type Deps struct {
 
 	Metrics metrics.Provider
 
-	// Timeout is configurable in site config.
-	Timeout time.Duration
```
Comment (evankanderson) [inline deps/deps.go:91]: This was never used, it turned out.

File: hugofs/files/classifier.go [hunk 3]
```diff
@@ -100,22 +100,13 @@ func ClassifyContentFile(filename string, open func() (afero.File, error)) Conte
...
 	}
-
-	if strings.HasPrefix(filename, "_index.") {
-		return ContentClassBranch
-	}
-
-	if strings.HasPrefix(filename, "index.") {
-		return ContentClassLeaf
-	}
-
```
Comment (evankanderson) [inline hugofs/files/classifier.go:118]: This determination moves into the only caller, `filter_fs.go`.

File: hugolib/fileInfo.go [hunk 4]
```diff
@@ -89,7 +89,18 @@ const (
...
+		switch fim.Meta().Classifier() {
+		case files.ContentClassBranch:
+			return bundleBranch, true
+		case files.ContentClassLeaf:
+			return bundleLeaf, true
+		default:
+			return bundleNot, false
+		}
+	}
+	// Fallback path, generally used for tests
```
Comment (evankanderson) [inline hugolib/fileInfo.go:103]: It turns out this fallback isn't needed, removing.

**Comment by evankanderson** [comment]
(I'll squash commits after I have the docs in)

**Comment by bep** [comment]
Hey,

Thanks for this. The code looks good, but I have 2 general comments. Regarding this I'm mostly convinced that:

1. Any config for this should live on the mount config (as in: overridden value for "_index" for a given mount)
2. This should be normalized when  I call `.Path`. That sentence is obviously simplified, but there will be situations where 2 content files take up the same place and needs some merge, e.g.: (content1)/posts/README.md and (content2)/posts/_index.md. This will be even more relevant/complicated once we start talking about non-markdown content. Doing this will also tell us if this "if _index" logic is too much spread out/duplicated.

**Comment by evankanderson** [comment]
For 1, let me take a look at where modules tie in. To be honest, I didn't find the documentation on modules at all until you gave me that keyword. From a quick look, it looks like a reasonable suggestion.

For 2, see the changes in `filter_fs.go`, where the first (highest priority) prefix in the list is promoted to be a bundle, and other files that may be present are treated as regular content pages. I _think_ that this allows a migration for sites (like my own) that are currently using kludges like `_index.md` files that just contain a `{{%readfile file="README.md"%}}` as content (this breaks things like use of shortcodes inside the included README). You can see this working (I think) in the e2e test here:
https://github.com/gohugoio/hugo/blob/f859d32d175dbff1693d0e43b02721a8a2fba949/hugolib/hugo_sites_build_test.go#L863
```
branchBundlePrefix = ["README", "test"]
...
		"sub2/test.md", `---
title: lower priority, just content
---
# Test content`,
....
	b.AssertFileContent("public/sub2/test/index.html", "Test content")
```

**Comment by evankanderson** [comment]
I'm assuming that the `branchBundlePrefix` option would only apply to content mounts?

**Comment by evankanderson** [comment]
For #1, it looks like the best option is to pass the list of files via FileMeta from the composite filesystem in `rootmapping_fs.go` to the filter list in `filter_fs.go`. This seems like it ha a few problems:

1. It *should* be the case that the FileMeta is the same for all the files in `applyMeta`, but there's nothing that guarantees that is true. This could potentially produce strange results based on mounts in subdirs.
2. I'm not sure about the layering; it seems strange to have the content determination split across the layers like this.

What would you think of having the RootMappingFs include a "branch score" on each file in a dir, and having `applyMeta` look at the top-scoring files to determine if the directory has a branch or leaf index file?

**Comment by bep** [comment]
>It should be the case that the FileMeta is the same for all the files in applyMeta, but there's nothing that guarantees that is true. 

I'm not sure I'm following:

* Every file inside /content will be member of a file mount
* That file mount can have it's `branchName="README"` set or not.

The above should be enough to determine if `Name()` (or whatever FileInfo method) should return `_index.foo` or not.

**Pull request closed without merging**